nemo/collections/asr/models/classification_models.py

```diff
@@ -53,23 +54,199 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
             self.local_rank = trainer.local_rank
 
         super().__init__(cfg=cfg, trainer=trainer)
-        self._update_decoder_config(self._cfg.decoder)
 
-        self.preprocessor = EncDecClassificationModel.from_config_dict(self._cfg.preprocessor)
-        self.encoder = EncDecClassificationModel.from_config_dict(self._cfg.encoder)
-        self.decoder = EncDecClassificationModel.from_config_dict(self._cfg.decoder)
-        self.loss = CrossEntropyLoss()
+        self.is_regression_task = self._cfg.get('is_regression_task', False)
         if hasattr(self._cfg, 'spec_augment') and self._cfg.spec_augment is not None:
-            self.spec_augmentation = EncDecClassificationModel.from_config_dict(self._cfg.spec_augment)
+            self.spec_augmentation = ASRModel.from_config_dict(self._cfg.spec_augment)
         else:
             self.spec_augmentation = None
         if hasattr(self._cfg, 'crop_or_pad_augment') and self._cfg.crop_or_pad_augment is not None:
-            self.crop_or_pad = EncDecClassificationModel.from_config_dict(self._cfg.crop_or_pad_augment)
+            self.crop_or_pad = ASRModel.from_config_dict(self._cfg.crop_or_pad_augment)
         else:
             self.crop_or_pad = None
 
-        # Setup metric objects
-        self._accuracy = TopKClassificationAccuracy(dist_sync_on_step=True)
+        self.preprocessor = self._setup_preprocessor()
+        self.encoder = self._setup_encoder()
+        self.decoder = self._setup_decoder()
+        self.loss = self._setup_loss()
+        self._setup_metrics()
+
+    @abstractmethod
+    def _setup_preprocessor(self):
+        """
+        Setup preprocessor for audio data
+        Returns: Preprocessor
+
+        """
+        pass
+
+    @abstractmethod
+    def _setup_encoder(self):
+        """
+        Setup encoder for the Encoder-Decoder network
+        Returns: Encoder
+        """
+        pass
+
+    @abstractmethod
+    def _setup_decoder(self):
+        """
+        Setup decoder for the Encoder-Decoder network
+        Returns: Decoder
+        """
+        pass
+
+    @abstractmethod
+    def _setup_loss(self):
+        """
+        Setup loss function for training
+        Returns: Loss function
+
+        """
+        pass
+
+    @abstractmethod
+    def _setup_metrics(self):
+        """
+        Setup metrics to be tracked in addition to loss
+        Returns: void
+
+        """
+        pass
+
+    @property
+    def input_types(self) -> Optional[Dict[str, NeuralType]]:
+        if hasattr(self.preprocessor, '_sample_rate'):
+            audio_eltype = AudioSignal(freq=self.preprocessor._sample_rate)
+        else:
+            audio_eltype = AudioSignal()
+        return {
+            "input_signal": NeuralType(('B', 'T'), audio_eltype),
+            "input_signal_length": NeuralType(tuple('B'), LengthsType()),
+        }
+
```

Comment:
Can you make the `output_types` attribute here abstract, so  that subclasses are forced to implement it (even a return None). Theres a high chance newer subclasses of this will forget to add the `output_types` since `input_types` is already defined.